modules/core/src/main/java/org/apache/ignite/spi/communication/tcp/InboundConnectionHandler.java

```diff
@@ -0,0 +1,930 @@
+                        }
+                    }
+                    else
+                        outDesc.onNodeLeft();
+                }
+            }
+
+            CommunicationListener<Message> lsnr0 = lsnr;
+
+            if (lsnr0 != null)
+                lsnr0.onDisconnected(id);
+        }
+    }
+
+    /**
+     * @param stopping New stopping flag (set to  when SPI gets stopping signal).
+     */
+    public void stop() {
+        this.stopping = true;
+    }
+
+    /**
+     * @param ses Session.
+     * @param msg Message.
+     */
+    private void onFirstMessage(final GridNioSession ses, Message msg) {
+        UUID sndId;
+
+        ConnectionKey connKey;
+
+        if (msg instanceof NodeIdMessage) {
+            sndId = U.bytesToUuid(((NodeIdMessage)msg).nodeIdBytes(), 0);
+            connKey = new ConnectionKey(sndId, 0, -1);
+        }
+        else {
+            assert msg instanceof HandshakeMessage : msg;
+
+            HandshakeMessage msg0 = (HandshakeMessage)msg;
+
+            sndId = ((HandshakeMessage)msg).nodeId();
+            connKey = new ConnectionKey(sndId, msg0.connectionIndex(), msg0.connectCount());
+        }
+
+        if (log.isDebugEnabled())
+            log.debug("Remote node ID received: " + sndId);
+
+        final ClusterNode rmtNode = nodeGetter.apply(sndId);
+
+        if (rmtNode == null) {
+            DiscoverySpi discoverySpi = igniteExSupplier.get().configuration().getDiscoverySpi();
+
+            boolean unknownNode = true;
+
+            if (discoverySpi instanceof TcpDiscoverySpi) {
+                TcpDiscoverySpi tcpDiscoverySpi = (TcpDiscoverySpi)discoverySpi;
+
+                ClusterNode node0 = tcpDiscoverySpi.getNode0(sndId);
+
+                if (node0 != null) {
+                    assert node0.isClient() : node0;
+
+                    if (node0.version().compareTo(VERSION_SINCE_CLIENT_COULD_WAIT_TO_CONNECT) >= 0)
+                        unknownNode = false;
+                }
+            }
+            else if (discoverySpi instanceof IgniteDiscoverySpi)
+                unknownNode = !((IgniteDiscoverySpi)discoverySpi).knownNode(sndId);
+
+            if (unknownNode) {
+                U.warn(log, "Close incoming connection, unknown node [nodeId=" + sndId + ", ses=" + ses + ']');
+
+                ses.send(new RecoveryLastReceivedMessage(UNKNOWN_NODE)).listen(fut -> ses.close());
+            }
+            else
+                ses.send(new RecoveryLastReceivedMessage(NEED_WAIT)).listen(fut -> ses.close());
+
+            return;
+        }
+
+        ses.addMeta(CONSISTENT_ID_META, rmtNode.consistentId());
+
+        final ConnectionKey old = ses.addMeta(CONN_IDX_META, connKey);
+
+        assert old == null;
+
+        ClusterNode locNode = locNodeSupplier.get();
+
+        if (ses.remoteAddress() == null)
+            return;
+
+        assert msg instanceof HandshakeMessage : msg;
```

Comment:
It is unclear where NodeIdMessage processing ends up.
It would be nice to have a comment or to assert some invariant on this.